File: typescript/test/scale.test.ts
import { readFileSync } from "node:fs";
import { resolve } from "node:path";
import { describe, expect, it } from "vitest";

import { decodeBytes, decodeCompact, encodeCompact } from "../src/scale.js";

const fixture = JSON.parse(
  readFileSync(resolve(__dirname, "../../e2e/scale-vectors.json"), "utf-8"),
);

function unhex(s: string): Uint8Array {
  return Uint8Array.from(Buffer.from(s.replace(/^0x/, ""), "hex"));
}

describe("scale compact codec", () => {
  it("encodes and decodes mode 0 lower bound", () => {
    expect(encodeCompact(0n)).toEqual(new Uint8Array([0x00]));
    expect(decodeCompact(new Uint8Array([0x00]))).toEqual([0n, 1]);
  });

  it("encodes and decodes mode 0 upper bound", () => {
    expect(encodeCompact(63n)).toEqual(new Uint8Array([0xfc]));
    expect(decodeCompact(new Uint8Array([0xfc]))).toEqual([63n, 1]);
  });

  it("encodes and decodes mode 1 lower bound", () => {
    expect(encodeCompact(64n)).toEqual(new Uint8Array([0x01, 0x01]));
    expect(decodeCompact(new Uint8Array([0x01, 0x01]))).toEqual([64n, 2]);
  });

  it("encodes and decodes mode 1 upper bound", () => {
    const encoded = encodeCompact(16_383n);
    expect(decodeCompact(encoded)).toEqual([16_383n, 2]);
  });

  it("encodes and decodes mode 2 lower bound", () => {
    const encoded = encodeCompact(16_384n);
    expect(decodeCompact(encoded)).toEqual([16_384n, 4]);
  });

  it("encodes and decodes mode 2 upper bound", () => {
    const encoded = encodeCompact((1n << 30n) - 1n);
    expect(decodeCompact(encoded)).toEqual([(1n << 30n) - 1n, 4]);
  });

  it("encodes and decodes big-int mode at 2^30", () => {
    const v = 1n << 30n;
    const decoded = decodeCompact(encodeCompact(v));
    expect(decoded?.[0]).toEqual(v);
  });

  it("encodes and decodes big-int mode at u64 max", () => {
    const v = (1n << 64n) - 1n;
    const decoded = decodeCompact(encodeCompact(v));
    expect(decoded?.[0]).toEqual(v);
  });

  it("round trips across all probe values", () => {
    const probes: bigint[] = [
      0n,
      1n,
      63n,
      64n,
      100n,
      16_383n,
      16_384n,
      1n << 20n,
      (1n << 30n) - 1n,
      1n << 30n,
      1n << 32n,
      (1n << 64n) - 1n,
    ];
    for (const v of probes) {
      const decoded = decodeCompact(encodeCompact(v));
      expect(decoded?.[0]).toEqual(v);
    }
  });

  it("returns null on empty input", () => {
    expect(decodeCompact(new Uint8Array())).toBeNull();
  });

  it("returns null on truncated mode 1", () => {
    expect(decodeCompact(new Uint8Array([0x01]))).toBeNull();
  });

  it("returns null on truncated mode 2", () => {
    expect(decodeCompact(new Uint8Array([0x02, 0x00, 0x00]))).toBeNull();
  });

  it("returns null on truncated big-int mode", () => {
    expect(decodeCompact(new Uint8Array([0x03, 0x01]))).toBeNull();
  });

  it("decodes a length-prefixed byte vector", () => {
    const wire = new Uint8Array([
      ...encodeCompact(5n),
      ...new TextEncoder().encode("hello"),
    ]);
    const result = decodeBytes(wire);
    expect(result).not.toBeNull();
    if (result === null) return;
    expect(new TextDecoder().decode(result[0])).toEqual("hello");
    expect(result[1]).toEqual(6);
  });

  it("returns null on truncated payload", () => {
    const wire = new Uint8Array([
      ...encodeCompact(10n),
      ...new TextEncoder().encode("only5"),
    ]);
    expect(decodeBytes(wire)).toBeNull();
  });

  it("matches the e2e scale-vectors fixture", () => {
    for (const c of fixture.compact) {
      const value = BigInt(c.value);
      const expected = unhex(c.encoded);
      expect(encodeCompact(value)).toEqual(expected);
      const decoded = decodeCompact(expected);
      expect(decoded).not.toBeNull();
      if (decoded === null) continue;
      expect(decoded[0]).toEqual(value);
      expect(decoded[1]).toEqual(c.consumed);
    }
  });
});

File: go/scale_test.go
package samp

import (
	"encoding/hex"
	"encoding/json"
	"math"
	"os"
	"path/filepath"
	"strconv"
	"testing"

	"github.com/stretchr/testify/require"
)

type scaleCompactCase struct {
	Value    string `json:"value"`
	Encoded  string `json:"encoded"`
	Consumed int    `json:"consumed"`
}

type scaleVectors struct {
	Compact []scaleCompactCase `json:"compact"`
}

func unhexBytes(t *testing.T, s string) []byte {
	t.Helper()
	if len(s) >= 2 && s[:2] == "0x" {
		s = s[2:]
	}
	b, err := hex.DecodeString(s)
	require.NoError(t, err)
	return b
}

func TestCompactModeZeroLowerBound(t *testing.T) {
	require.Equal(t, []byte{0x00}, EncodeCompact(0))
	v, n, err := DecodeCompact([]byte{0x00})
	require.NoError(t, err)
	require.Equal(t, uint64(0), v)
	require.Equal(t, 1, n)
}

func TestCompactModeZeroUpperBound(t *testing.T) {
	require.Equal(t, []byte{0xfc}, EncodeCompact(63))
	v, n, err := DecodeCompact([]byte{0xfc})
	require.NoError(t, err)
	require.Equal(t, uint64(63), v)
	require.Equal(t, 1, n)
}

func TestCompactModeOneLowerBound(t *testing.T) {
	require.Equal(t, []byte{0x01, 0x01}, EncodeCompact(64))
	v, n, err := DecodeCompact([]byte{0x01, 0x01})
	require.NoError(t, err)
	require.Equal(t, uint64(64), v)
	require.Equal(t, 2, n)
}

func TestCompactModeOneUpperBound(t *testing.T) {
	enc := EncodeCompact(16383)
	v, n, err := DecodeCompact(enc)
	require.NoError(t, err)
	require.Equal(t, uint64(16383), v)
	require.Equal(t, 2, n)
}

func TestCompactModeTwoLowerBound(t *testing.T) {
	enc := EncodeCompact(16384)
	v, n, err := DecodeCompact(enc)
	require.NoError(t, err)
	require.Equal(t, uint64(16384), v)
	require.Equal(t, 4, n)
}

func TestCompactModeTwoUpperBound(t *testing.T) {
	enc := EncodeCompact(1<<30 - 1)
	v, n, err := DecodeCompact(enc)
	require.NoError(t, err)
	require.Equal(t, uint64(1<<30-1), v)
	require.Equal(t, 4, n)
}

func TestCompactBigIntMode2Pow30(t *testing.T) {
	v, _, err := DecodeCompact(EncodeCompact(1 << 30))
	require.NoError(t, err)
	require.Equal(t, uint64(1<<30), v)
}

func TestCompactBigIntModeU64Max(t *testing.T) {
	v, _, err := DecodeCompact(EncodeCompact(math.MaxUint64))
	require.NoError(t, err)
	require.Equal(t, uint64(math.MaxUint64), v)
}

func TestCompactRoundTripAcrossAllModes(t *testing.T) {
	probes := []uint64{0, 1, 63, 64, 100, 16383, 16384, 1 << 20, (1 << 30) - 1, 1 << 30, 1 << 32, math.MaxUint64}
	for _, value := range probes {
		decoded, _, err := DecodeCompact(EncodeCompact(value))
		require.NoError(t, err)
		require.Equal(t, value, decoded)
	}
}

func TestDecodeCompactReturnsErrorOnEmptyInput(t *testing.T) {
	_, _, err := DecodeCompact([]byte{})
	require.ErrorIs(t, err, ErrInsufficientData)
}

func TestDecodeCompactReturnsErrorOnTruncatedTwoByteMode(t *testing.T) {
	_, _, err := DecodeCompact([]byte{0x01})
	require.ErrorIs(t, err, ErrInsufficientData)
}

func TestDecodeCompactReturnsErrorOnTruncatedFourByteMode(t *testing.T) {
	_, _, err := DecodeCompact([]byte{0x02, 0x00, 0x00})
	require.ErrorIs(t, err, ErrInsufficientData)
}

func TestDecodeCompactReturnsErrorOnTruncatedBigIntMode(t *testing.T) {
	_, _, err := DecodeCompact([]byte{0x03, 0x01})
	require.ErrorIs(t, err, ErrInsufficientData)
}

func TestDecodeBytesExtractsPayloadAfterCompactLength(t *testing.T) {
	wire := append(EncodeCompact(5), []byte("hello")...)
	payload, consumed, err := DecodeBytes(wire)
	require.NoError(t, err)
	require.Equal(t, []byte("hello"), payload)
	require.Equal(t, 6, consumed)
}

func TestDecodeBytesReturnsErrorWhenPayloadTruncated(t *testing.T) {
	wire := append(EncodeCompact(10), []byte("only5")...)
	_, _, err := DecodeBytes(wire)
	require.ErrorIs(t, err, ErrInsufficientData)
}

func TestDecodeCompactBigIntModeTruncatedPayload(t *testing.T) {
	// BigInt mode: low 2 bits = 0b11, upper 6 bits encode (n-4) where n is byte count.
	// 0x13 = 0b00010011 -> mode=0b11, n = (0x13>>2)+4 = 4+4 = 8 bytes follow.
	// But we only provide 3 bytes after the prefix.
	_, _, err := DecodeCompact([]byte{0x13, 0x01, 0x02, 0x03})
	require.ErrorIs(t, err, ErrInsufficientData)
}

func TestDecodeBytesCompactOverflow(t *testing.T) {
	// Compact length that would overflow int when computing end offset.
	// A compact-encoded very large value. We use BigInt mode with n=8 (max).
	// 0x13 means 8 bytes follow. Set them to large values.
	data := []byte{0x13, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0x7F}
	_, _, err := DecodeBytes(data)
	require.ErrorIs(t, err, ErrInsufficientData)
}

func TestDecodeBytesForwardDecodeCompactError(t *testing.T) {
	// Empty input should forward the DecodeCompact error.
	_, _, err := DecodeBytes([]byte{})
	require.ErrorIs(t, err, ErrInsufficientData)
}

func TestMatchesE2EScaleVectorsFixture(t *testing.T) {
	path := filepath.Join("..", "e2e", "scale-vectors.json")
	raw, err := os.ReadFile(path)
	require.NoError(t, err)

	var vectors scaleVectors
	require.NoError(t, json.Unmarshal(raw, &vectors))

	for _, c := range vectors.Compact {
		value, err := strconv.ParseUint(c.Value, 10, 64)
		require.NoError(t, err)

		expected := unhexBytes(t, c.Encoded)
		require.Equal(t, expected, EncodeCompact(value))

		decoded, consumed, err := DecodeCompact(expected)
		require.NoError(t, err)
		require.Equal(t, value, decoded)
		require.Equal(t, c.Consumed, consumed)
	}
}
